core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala

```diff
@@ -91,6 +91,50 @@ private[spark] object ClosureCleaner extends Logging {
     (seen - obj.getClass).toList
   }
 
+  /** Initializes the accessed fields for outer classes and their super classes. */
+  private def initAccessedFields(
+      accessedFields: Map[Class[_], Set[String]],
+      outerClasses: Seq[Class[_]]): Unit = {
+    for (cls <- outerClasses) {
+      var currentClass = cls
+      do {
+        accessedFields(currentClass) = Set.empty[String]
+        currentClass = currentClass.getSuperclass()
+      } while (currentClass != null)
+    }
+  }
+
+  /** Sets accessed fields for given class in clone object based on given object. */
+  private def setAccessedFields(
+      outerClass: Class[_],
+      clone: AnyRef,
+      obj: AnyRef,
+      accessedFields: Map[Class[_], Set[String]]): Unit = {
+    for (fieldName <- accessedFields(outerClass)) {
+      val field = outerClass.getDeclaredField(fieldName)
+      field.setAccessible(true)
+      val value = field.get(obj)
+      field.set(clone, value)
+    }
+  }
+
+  /** Clones a given object and sets accessed fields in cloned object. */
+  private def cloneAndSetFields(
+      parent: AnyRef,
+      obj: AnyRef,
+      outerClass: Class[_],
+      accessedFields: Map[Class[_], Set[String]]): AnyRef = {
+    val clone = instantiateClass(outerClass, parent)
+
+    var currentClass = outerClass
+    do {
+      setAccessedFields(currentClass, clone, obj, accessedFields)
```

Comment:
Assume we have class `A` and `B` having the same parent class `P`. `P` has 2 fields `a` and `b`. The closure accessed `A.a` and `B.b`, so when we clone `A` object, we should only set field `a`, when we clone `B` object, we should only set field `b`. However here seems we set field `a` and `b` for `A` and `B` object, which is sub-optimal.

Reply:
Seems this is also a issue for the `outerClasses`, maybe I missed something...

Reply:
Seems that is true. For a closure that only accessed `A.a`, we clone the whole `A` object which contains both `a` and `b` fields. This is the fact in current `ClosureCleaner`.

Reply:
As this is not a regression, IIUC, will it block this change?

Reply:
yea let's leave it